src/mlabns/mlabns_impl.hpp

```diff
@@ -0,0 +1,114 @@
+// information on the copying conditions.
+
+#include "src/ext/json/src/json.hpp"
+#include <measurement_kit/common.hpp>
+#include <measurement_kit/http.hpp>
+#include <measurement_kit/mlabns.hpp>
+#include <regex>
+#include <string>
+#include <utility>
+
+using json = nlohmann::json;
+
+namespace mk {
+namespace mlabns {
+
+static inline ErrorOr<std::string> as_query(Settings &settings) {
+    std::string query;
+    std::string policy = settings.get<std::string>("mlabns/policy", "");
+    std::string metro = settings.get<std::string>("mlabns/metro", "");
+    std::string address_family =
+        settings.get<std::string>("mlabns/address_family", "");
+    if (policy == "" && metro == "" && address_family == "") {
+        return query;
+    }
+    if (policy != "") {
+        if (policy != "geo" && policy != "random" && policy != "metro" &&
+            policy != "country") {
+            return InvalidPolicyError();
+        }
+        if (query != "") {
+            query += "&";
+        }
+        query += "policy=" + policy;
+    }
+    if (metro != "") {
+        std::regex valid_metro("^[a-z]{3}$");
+        if (!std::regex_match(metro, valid_metro)) {
+            return InvalidMetroError();
+        }
+        if (query != "") {
+            query += "&";
+        }
+        query += "metro=" + metro;
+    }
+    if (address_family != "") {
+        if (address_family != "ipv4" && address_family != "ipv6") {
+            return InvalidAddressFamilyError();
+        }
+        if (query != "") {
+            query += "&";
+        }
+        query += "address_family=" + address_family;
+    }
+    query = "?" + query;
+    return query;
+}
+
+template <MK_MOCK_NAMESPACE(http, get)>
+inline void query_impl(std::string tool, Callback<Error, Reply> callback,
+                        Settings settings, Var<Reactor> reactor,
+                        Var<Logger> logger) {
+    ErrorOr<std::string> query = as_query(settings);
+    if (!query) {
+        callback(query.as_error(), Reply());
+        return;
+    }
+    std::string url = "https://mlab-ns.appspot.com/";
+    std::regex valid_tool("^[a-z]+$");
+    if (!std::regex_match(tool, valid_tool)) {
+        callback(InvalidToolNameError(), Reply());
+        return;
+    }
+    url += tool;
+    url += *query;
+    logger->info("query mlabns for tool %s", tool.c_str());
+    logger->debug("mlabns url: %s", url.c_str());
+    http_get(url,
+        [callback, logger](Error error, http::Response response) {
+            if (error) {
+                callback(error, Reply());
+                return;
+            }
+            if (response.status_code != 200) {
+                callback(UnexpectedHttpStatusCodeError(), Reply());
+                return;
+            }
+            Reply reply;
+            try {
+                auto node = json::parse(response.body);
+                reply.city = node.at("city");
+                reply.url = node.at("url");
```

Comment:
I see that you have changed `node["city"]` in `node.at("city")`. As we discussed in private, I think this change is an improvement because the documentation of nlohmann/json clearly points out that `at()` throws if the element is missing, while it is not as clear on that point for `[]` (even though currently `[]` throws as well). But, at the same time, I think that in this case it would probably have been better to separate the refactoring and further changes (namely clang-format and applying this other change from `[]` to `at()`), to simplify the job of who is reviewing the diff.